golem-worker-executor-base/src/worker/mod.rs

```diff
@@ -873,6 +899,105 @@ impl<Ctx: WorkerCtx> Worker<Ctx> {
         Ok(())
     }
 
+    /// Reverts the worker to a previous state, selected by either the last oplog index to keep
+    /// or the number of invocations to drop.
+    ///
+    /// The revert operations is implemented by inserting a special oplog entry that
+    /// extends the worker's deleted oplog regions, skipping entries from the end of the oplog.
+    pub async fn revert(&self, target: RevertWorkerTarget) -> Result<(), GolemError> {
+        match target {
+            RevertWorkerTarget::RevertToOplogIndex(target) => {
+                self.revert_to_last_oplog_index(target.last_oplog_index)
+                    .await
+            }
+            RevertWorkerTarget::RevertLastInvocations(target) => {
+                if let Some(last_oplog_index) = self
+                    .find_nth_invocation_from_end(target.number_of_invocations as usize)
+                    .await
+                {
+                    self.revert_to_last_oplog_index(last_oplog_index.previous())
+                        .await
+                } else {
+                    Err(GolemError::invalid_request(format!(
+                        "Could not find {} invocations to revert",
+                        target.number_of_invocations
+                    )))
+                }
+            }
+        }
+    }
+
+    /// Starting from the end of the oplog, find the Nth ExportedFunctionInvoked entry's index.
+    async fn find_nth_invocation_from_end(&self, n: usize) -> Option<OplogIndex> {
```

Comment:
How will should this behave if we have updates / memory increases in the oplog region that is getting deleted?
We might want to delete region belonging to invocations, but keep other regions that belong to actions that users performed.

Reply:
This feature reverts everything to a given point in the oplog. Including worker updates, pending invocations, etc. I think this is what most users would expect. We could also have some more advanced "history editing" feature where you just remove random invocations and leave the rest of the history as is. I think such a feature could go into the debugging service later.